### ISSUE TITLE
Generic function expecting type parameter can't figure out type when Literal has only one parameter

### ISSUE THREAD

**Issue by joanofxyz** [microsoft/pyright#7870]
When calling a function that expects a generic `type` as a parameter with a `Literal` type, I get the following error when the `Literal` only has a single value. I can't understand if this is intended behaviour but I haven't been able to wrap my head around why this would work only when the `Literal` definition has multiple parameters  

```
Argument of type "Single" cannot be assigned to parameter "expected_type" of type "type[T@fn]" in function "fn"
  Type "type[Literal]" cannot be assigned to type "type[str]" [reportArgumentType]
```

<img width="1059" alt="image" src="https://github.com/microsoft/pyright/assets/6023880/942a498a-64af-49f0-afa7-9000c2bf69d4">

```python
from typing import Literal, TypeVar, get_args, get_origin

T = TypeVar("T")
Single = Literal["A"]
Multiple = Literal["A", "B"]


def fn(value: T, expected_type: type[T]):
    if get_origin(expected_type) == Literal:
        assert value in get_args(expected_type), f"value must be of type {expected_type}"
    else:
        assert isinstance(value, expected_type), f"value must be of type {expected_type}"


fn("A", Multiple)
fn("A", Single)
```

running pyright 1.1.357 as cli lsp for NVIM v0.9.5

**Comment by erictraut** [comment]
`Literal` is meant to be used in a type expression (annotation), but you're using it here in a value expression. When used in a type expression, `Literal["A"]` has a well-defined meaning to a static type checker as defined in the [Python typing spec](https://typing.readthedocs.io/en/latest/spec/literal.html). When used in a value expression, `Literal` has no defined meaning to a static type checker. It is effectively just an object of some sort. It is definitely not compatible with `type[T]`, so pyright is correct in generating an error here. For details about `type[T]` and how it is interpreted, see [this section](https://typing.readthedocs.io/en/latest/spec/special-types.html#type) of the typing spec.

Pyright should probably also generate an error for the expression `fn("A", Multiple)`. I'll look at whether that straightforward to do.

In any case, your code is not using the static type system properly. You'll need to change your code if you want it to type check.

You may be interested in [this thread](https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435/9) which is associated with a [draft PEP](https://docs.google.com/document/d/1PRvl3uKE-BxvmyFO3Ic4fZDpgW2fOU58aYh3LJ3zLpk/edit) that would support your use case.

**Comment by erictraut** [comment]
In the next release, pyright will properly flag `fn("A", Multiple)` as an error.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.362, which I just published.

**Comment by joanofxyz** [comment]
thank you for the thorough explanation!
